wandb/sdk/wandb_settings.py

```diff
@@ -223,925 +146,1015 @@ class SettingsConsole(enum.Enum):
+                source=Source.WORKSPACE,
             )
-        if self.run_id:
-            format_dict["run_id"] = self.run_id
-        format_dict["run_mode"] = "offline-run" if self._offline else "run"
-        format_dict["proc"] = os.getpid()
-        # TODO(cling): hack to make sure we read from local settings
-        #              this is wrong if the run_dir changes later
-        format_dict["wandb_dir"] = self.wandb_dir or "wandb"
-
-        path_items: List[str] = []
-        for p in path:
-            part = self._path_convert_part(p, format_dict)
-            if part is None:
-                return None
-            path_items += part
-        converted_path = os.path.join(*path_items)
-        converted_path = os.path.expanduser(converted_path)
-        return converted_path
-
-    # def _clear_early_logger(self) -> None:
-    #     # TODO(jhr): this is a hack
-    #     object.__setattr__(self, "_Settings__early_logger", None)
-
-    def _setup(self, kwargs: Any) -> None:
-        for k, v in six.iteritems(kwargs):
-            if k not in self._unsaved_keys:
-                object.__setattr__(self, k, v)
-
-    def __copy__(self) -> "Settings":
-        """Copy (note that the copied object will not be frozen)."""
-        s = Settings()
-        s._apply_settings(self)
-        return s
-
-    def duplicate(self) -> "Settings":
-        return copy.copy(self)
-
-    def _check_invalid(self, k: str, v: Any) -> None:
-        if v is None:
-            return
-        f = getattr(self, "_validate_" + k, None)
-        if not f or not callable(f):
-            return
-        invalid = f(v)
-        if invalid:
-            raise UsageError("Settings field `{}`: {}".format(k, invalid))
-
-    def _perform_preprocess(self, k: str, v: Any) -> Optional[Any]:
-        f = getattr(self, "_preprocess_" + k, None)
-        if not f or not callable(f):
-            return v
-        else:
-            return f(v)
 
-    def _update(
-        self,
-        __d: Dict[str, Any] = None,
-        _source: Optional[int] = None,
-        _override: Optional[int] = None,
-        **kwargs: Any,
+    def apply_env_vars(
+        self, environ: Mapping[str, Any], _logger: Optional[_EarlyLogger] = None,
     ) -> None:
-        if self.__frozen and (__d or kwargs):
-            raise TypeError("Settings object is frozen")
-        d = __d or dict()
-        data = {}
-        for check in d, kwargs:
-            for k in six.viewkeys(check):
-                if k not in self.__dict__:
-                    raise KeyError(k)
-                v = self._perform_preprocess(k, check[k])
-                self._check_invalid(k, v)
-                data[k] = v
-        for k, v in six.iteritems(data):
-            if v is None:
-                continue
-            if self._priority_failed(k, source=_source, override=_override):
+        env_prefix: str = "WANDB_"
+        special_env_var_names = {
```

Comment:
Would be good to add a note about what makes these specials and when a dev should add something to this list.